core/plugin_api.rs

```diff
@@ -20,4 +22,15 @@ pub type DispatchOpFn =
 
 pub trait Interface {
   fn register_op(&mut self, name: &str, dispatcher: DispatchOpFn) -> OpId;
+  fn resource_table(&mut self) -> &mut dyn ResourceTable;
+}
+
+/// Similar to `struct ResourceTable` for normal ops, but uses dynamic dispatch
+/// instead of type parameters for its methods.
+pub trait ResourceTable {
+  fn add(&mut self, name: &str, resource: Box<dyn Resource>) -> ResourceId;
+  fn close(&mut self, rid: ResourceId) -> Option<()>;
+  fn get(&self, rid: ResourceId) -> Option<&dyn Resource>;
+  fn get_mut(&mut self, rid: ResourceId) -> Option<&mut dyn Resource>;
+  fn has(&self, rid: ResourceId) -> bool;
```

Comment:
This is unfortunate. Is there no way to avoid this?

Reply:
Avoid what?

Reply:
Avoid adding this trait - it's a complication - there's only ever one resource table type.

Reply:
Not really I'm afraid. It's here for the same reason as the the other stuff in plugin_api -- to make dynamically imported plugins not crash when they're dropped.

I think I disagree with you that this is unfortunate though. I think it provides a clean delineation between the public API and our overall pretty messy implementation. We should move it to it's own crate at some point so it can provide a stable "plugin api" snapshot; since this is a trait it would not break as long as the trait definition stays the same, even if core moves on and changes APIs.

Also, you don't have to use it when you're writing a statically linked plugin.

Reply:
How do plugins execute javascript? It seems like the isolate was removed?

Reply:
The isolate was never exposed... And they're not supposed to touch v8?
But if we wanted to add that we could.

Reply:
This would still cause the same problem as using structs in the plugin api. I think that it would be best to avoid this. It would also make it hard to build a plugin api that isn't dependent on the rest of core.

Reply:
This would still cause the same problem as using structs in the plugin api. I think that it would be best to avoid this. It would also make it hard to build a plugin api that isn't dependent on the rest of core.

Reply:
Structs as such are not the problem though -- it's type parameters, closures, things that get monophormized and end up in the plugin lib that causes problems. ZeroCopyBuf is a struct and we expose it without any issues.

That doesn't mean I think that exposing CoreIsolate makes a lot of sense. That struct is full of internal stuff that we really ought to clean up some day. Definitely not something you want to freeze into a (future stable) API.
The only things that I can imagine would matter to people is the v8::Isolate and v8::Global, which we could just as easily expose directly on `Interface`.